Add trivial timeout handler to use with client channels

Spikhalskiy opened the pull request:
Motivation:

Writing client with netty everybody want to catch and control a situation of request timeout, or socket timeout at least. Now there are no suitable handlers in netty standard distribution for this. IdleStateHandler is suitable only for server sockets.

Modifications:

Add an extendable handler that makes it possible to mark write() call as a start point for timeout task and cancel this task on the first read from a channel (response received). “First read” policy could be overridden with other logic, for example, read of LastHttpContent message.

Result:

Users that implement own clients based on netty have trivial read timeout handler in standard netty classes.

netkins commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,203 @@
+     * @param msg message of related {@link #channelRead} invocation
+     * @return true if timeout should be cancelled by receiving this read, false otherwise
+     */
+    protected boolean timeoutShouldBeCancelled(ChannelHandlerContext ctx, Object msg) {
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused method parameter "ctx". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)

netkins replied:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused method parameter "ctx". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)

normanmaurer replied:
Add `@Supresswarnings("unused")`

normanmaurer commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,203 @@
+     */
+    public ReadAfterWriteTimeoutHandler(long timeout, TimeUnit unit, boolean closeOnTimeout) {
+        if (timeout <= 0) {
+            timeoutNanos = 0;
```
just throw if timeout is not positive ?

Spikhalskiy replied:
Made it consistent with ```io.netty.handler.timeout.WriteTimeoutHandler```. I could put throw in both places.

normanmaurer commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,203 @@
+        return true;
+    }
+
+    private void init(final ChannelHandlerContext ctx) {
```
seems like this code is only called in `handlerAdded`... Why not just move in there and remove all the null checks ?

normanmaurer commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,203 @@
+    /**
+     * Is called when a read timeout was detected.
+     */
+    private void readTimedOut(ChannelHandlerContext ctx) throws Exception {
```
mark protected so people can override it ?

normanmaurer commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,203 @@
+    /**
+     * This method is called after receiving each {@link #channelRead} and should define if this read should trigger
+     * timeout cancelling. Default implementation returns {@code true}.
+     * @param ctx channel handler context of related {@link #channelRead} invocation
```
{@link ChannelHandlerContext}

normanmaurer commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,203 @@
+
+    private long readFinishTime;
+    private long writeFinishTime;
+    private volatile boolean removed;
```
move volatile field below non volatile

normanmaurer commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,203 @@
+ * This handler supports trivial emulation of soTimeout for clients, measure time between
+ * last <code>write()</code> method invocation and first proceed <code>channelRead()</code>
+ *
+ * NB: Timeout measurement in default implementation
```
NB ?

Spikhalskiy replied:
"Nota bene" Could just delete.

Scottmitch replied:
+1 for removing `NB:`

normanmaurer commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,203 @@
+ * of extremely slow or dead reads.
+ *
+ */
+public class ReadAfterWriteTimeoutHandler extends ChannelDuplexHandler {
```
Please add some tests.

Spikhalskiy replied:
Looks like there is no tests for *TimeoutHandler or IdleStateHandler now in netty - will figure out and prepare test for this class.

NiteshKant commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,203 @@
+        }
+    }
+
+    private void initialize(ChannelHandlerContext ctx) {
```
This is only called from `init` so move it there?

NiteshKant commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,203 @@
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+        ctx.write(msg, promise.addListener(channelFutureListener));
```
Looks to me that we are adding the same listener for all writes and re-scheduling timeout on all write completions?

Spikhalskiy replied:
Yes. As a result, we have one scheduled task on last executed write().

NiteshKant replied:
Hmm, looks to me that if we have two writes then both of them will add the listener and on completion of both the writes, this code in `channelFutureListener` will be invoked:

```java
if (future.isSuccess()) {
                        if (removed) {
                            return;
                        }

                        writeFinishTime = System.nanoTime();
                        //init readFinishTime to writeFinishTime - readFinishTime > 0 be true if we don't get response
                      readFinishTime = writeFinishTime - 1;
                        if (timeoutNanos > 0) {
                            timeout = ctx.executor().schedule(readTimeoutTask, timeoutNanos, TimeUnit.NANOSECONDS);
                        }
                    }
```

which will schedule the timeout task twice, what am I missing?

Spikhalskiy replied:
Yep, you are right. Added ```disable()``` call to the listener before scheduling new timeout. https://github.com/netty/netty/pull/5484/files#diff-6a2c899583a1eae50e09ae517f50121aR94 Thanks for catching. WDYT about this fix?

NiteshKant commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,203 @@
+        }
+
+        writeFinishTime = System.nanoTime();
+        readFinishTime = writeFinishTime - 1;
```
Can you explain why `readFinishTime` is one nano second less than `writeFinishTime`?

Spikhalskiy replied:
Yes, if we don't have response - writeFinishTime - readFinishTime > 0 would be true after this initialization.

NiteshKant commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,203 @@
+
+        writeFinishTime = System.nanoTime();
+        readFinishTime = writeFinishTime - 1;
+        if (timeoutNanos > 0) {
```
This would always be true, as `timeoutNanos` is final and constructor makes sure that it is `> 0`

Spikhalskiy replied:
It could be zero. This behaviour copied from WriteTimeoutHandler. We should change them accordingly in that case.

NiteshKant commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,203 @@
+            if (closeOnTimeout) {
+                ctx.close();
+            }
+            closed = true;
```
May be just use `ctx.channel.isOpen()` rather than having a local flag to this handler?

Spikhalskiy replied:
Implemented it consistently with ReadTimeoutHandler or WriteTimeoutHandler, but yes, in IdleStateHandler we use this way - rework to ```ctx.channel.isOpen()```.

Spikhalskiy commented:
Addressed some comments, wrote answers to questionable once that should lead to refactoring in ReadTimeoutHandler and WriteTimeoutHandler to consistent behavior. Will prepare "functional"? test when I get a chance.

netkins commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,195 @@
+     */
+    @SuppressWarnings("unused")
+    protected boolean timeoutShouldBeCancelled(ChannelHandlerContext ctx, Object msg) {
+        return true;
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused method parameter "ctx". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)

netkins commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,195 @@
+     * @return true if timeout should be cancelled by receiving this read, false otherwise
+     */
+    @SuppressWarnings("unused")
+    protected boolean timeoutShouldBeCancelled(ChannelHandlerContext ctx, Object msg) {
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused method parameter "ctx". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)

Spikhalskiy commented:
@normanmaurer Hi, added tests and some configurable features to make things consistent. The current code works fine and currently using with http codec.

normanmaurer commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,219 @@
+            disable();
+        }
+        if (suppressReadAfterTimeout && timedOut) {
+            ctx.fireUserEventTriggered(ReadSuppressed.INSTANCE);
```
You will need to call `ReferenceCountUtil.release(msg)` to not leak.

Spikhalskiy replied:
Yep, thanks, fixed.

normanmaurer commented on handler/src/test/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandlerTest.java:
```diff
@@ -0,0 +1,345 @@
+package io.netty.handler.timeout;
```
license header

Spikhalskiy replied:
added

Spikhalskiy commented:
@normanmaurer addressed

netkins commented on handler/src/test/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandlerTest.java:
```diff
@@ -0,0 +1,360 @@
+
+        ChannelInboundHandler pongHandler = new ChannelInboundHandlerAdapter() {
+            public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                Thread.sleep(2 * TIMEOUT_MS);
```
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Remove this use of "Thread.sleep()". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2925)

netkins commented on handler/src/test/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandlerTest.java:
```diff
@@ -0,0 +1,360 @@
+            public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                serverReads.incrementAndGet();
+                if (serverReads.get() != MESSAGES_COUNT) {
+                    Thread.sleep(2 * TIMEOUT_MS);
```
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Remove this use of "Thread.sleep()". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2925)

netkins commented on handler/src/test/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandlerTest.java:
```diff
@@ -0,0 +1,360 @@
+        Channel client = newClient(server.localAddress(), readAfterWriteTimeoutHandler, trackingHandler);
+        try {
+            client.writeAndFlush(newOneMessage());
+            Thread.sleep(TIMEOUT_MS / 10);
```
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Remove this use of "Thread.sleep()". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2925)

netkins commented on handler/src/test/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandlerTest.java:
```diff
@@ -0,0 +1,360 @@
+                serverReads.incrementAndGet();
+                //response once only after MESSAGES_COUNT messages and sleeping
+                if (serverReads.get() == MESSAGES_COUNT) {
+                    Thread.sleep(2 * TIMEOUT_MS);
```
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Remove this use of "Thread.sleep()". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2925)

netkins commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,221 @@
+        }
+    }
+
+    public static final class ReadSuppressed {
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add a private constructor to hide the implicit public one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1118)

netkins commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,221 @@
+     * @param msg message of related {@link #channelRead} invocation
+     * @return true if timeout should be cancelled by receiving this read, false otherwise
+     */
+    protected boolean timeoutShouldBeCancelled(@SuppressWarnings("unused") ChannelHandlerContext ctx,
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused method parameter "ctx". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)

Scottmitch commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,223 @@
+
+/**
+ * This handler supports trivial emulation of soTimeout for clients, measure time between
+ * last <code>write()</code> method invocation and first proceed <code>channelRead()</code>
```
use `{@code}` instead of `<code>`?

Scottmitch commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,223 @@
+     * Creates a new instance.
+     *
+     * @param timeout
+     *        read timeout after last write to the channel
```
nit: put on the previous line.

same comment for other `@param` javadocs

Spikhalskiy replied:
it's just done with the same style as other javadocs in the ```io.netty.handler.timeout```package. Ok, would put on the same line.

Scottmitch commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,223 @@
+     * Also this addresses situation when we don't close connection after timeout ({@code closeOnTimeout == false}) and
+     * should skip next read which was received after timeout.
+     * User event {@link ReadSuppressed} will be thrown if case of read suppressing.
+     *
```
kill line

Scottmitch commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,223 @@
+ * if you have marker of last event - it's worth to override {@link #timeoutShouldBeCancelled} method,
+ * otherwise - use this handler in pair with reasonable {@link IdleStateHandler} to catch situations
+ * of extremely slow or dead reads.
+ *
```
kill line

Scottmitch commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,223 @@
+    }
+
+    /**
+     * This method is called after receiving each {@link #channelRead} and should define if this read should trigger
```
doesn't the `@link` need parameters types? `{@link #channelRead(ChannelHandlerContext, Object)}`

Scottmitch commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,223 @@
+                    //ensure that only the last write listener in a batch would setup one final timeout task
+                    //or nothing in a case of error
+                    disable();
+                    timedOut = false;
```
Can you explain the intended use case for `suppressReadAfterTimeout`? The combination of resetting `timedOut` here and `suppressReadAfterTimeout` may lead to confusing situations. Intermediate messages may be dropped and it may not be possible for the protocol to recover from this.

Spikhalskiy replied:
I personally use this handler with suppressReadAfterTimeout = true.

Usecase. We write a proxy. And we want to pass response back or pass 204 if no response got. We don't interested in response after timeout at all. It's actually quite often situation with having timeouts - you are interesting in a result in time and already not after that. So, suppressReadAfterTimeout = true would make it possible to simplify state machine in the main handler after ```ReadAfterWiteTimeoutHandler```. This handler gets reads only if there is no timeouts and has no ignore logic for reads after timeout.
I think that it's quite low lovel handling to be implemented in out-of-the box handler to make it possible to don't reimplement in the main business-logic handler.

Scottmitch replied:
How is writing more data correlated with the application wanting to read more data in this scenario? It seems like what you want to implement is a request/response based discard ... if the "current" request has timed out then discard all data from the corresponding response? The local endpoint could write a new request even if an outstanding request is timed out (e.g. in http pipelining could be used) ... but it seems like you still want to discard any data for the "current" response?

Scottmitch replied:
> I think this hander would not pair with pipelining correctly

That is my assumption as well. We need to make sure all HTTP handlers "support", or don't break if pipelining is in use.

Scottmitch commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,223 @@
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+        ctx.write(msg, promise.addListener(channelOnWriteListener));
```
`readTimeoutTask` is only executed once the write completes, but what if the peer is not reading and the write doesn't complete?

Spikhalskiy replied:
As with long reads, IdleStateHandler should be used, would add clarification to the class javadoc.

Scottmitch replied:
Can you clarify what you mean `as with long reads`?

It seems like the intended semantics of this handler are not being correctly enforced in this case. Can you explain why you recommend using a handler with different timeout semantics to account for this case, and how do you reconcile the differences in semantics between the two handlers when they are both in use?

Spikhalskiy replied:
Semantic of this handler:
You write a client and you need to make classical timeout semantic. You wrote everything successfully to a socket and want to get a result in 100ms from the server and pass it to the caller, or return "204" entity to the caller if no answer got in this timeout.

So, this handler doesn't cover situation with "write doesn't complete". It's to handle fair server response time between finishing write on the client side and reading a response. This is how it should work in my mind. Do you propose to extend semantic?

Scottmitch replied:
> So, this handler doesn't cover situation with "write doesn't complete".

Why is it necessary to make this distinction? Is the user able to distinguish the difference, and what do you expect the user to do in each case? "write doesn't complete" may be because the server is not reading fast enough, or it is not reading at all, or it is dead ... I don't understand why it is desirable to ignore this.

Scottmitch commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,223 @@
+                        writeFinishTime = System.nanoTime();
+                        //init readFinishTime to writeFinishTime - readFinishTime > 0 be true if we don't get response
+                        readFinishTime = writeFinishTime - 1;
+                        if (timeoutNanos > 0) {
```
what does this handler do if `timeoutNanos <= 0`? Should we even allow this?

Spikhalskiy replied:
```
if (timeout <= 0) {
            timeoutNanos = 0;
        } else {
            timeoutNanos = Math.max(unit.toNanos(timeout), MIN_TIMEOUT_NANOS);
        }
```

This is done in a consistent manner with the code in other handlers in ```io.netty.handler.timeout```

Scottmitch commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,223 @@
+                        }
+                        writeFinishTime = System.nanoTime();
+                        //init readFinishTime to writeFinishTime - readFinishTime > 0 be true if we don't get response
+                        readFinishTime = writeFinishTime - 1;
```
Is it necessary to use `nanoTime()` for `readFinishTime` and `writeFinishTime`? It seems like the timeout detection is more based off of if we have received a read since the last time we wrote ... can this be done with a boolean `sawRead` that is set in `channelRead` (maybe based upon something like `timeoutShouldBeCancelled`)  and reset in `channelWrite`?  This handler could be extended and then only reset the `sawRead` for example if the `last message` is detected.

Spikhalskiy replied:
Good idea, thanks, it really looks simpler. Reworked.

Spikhalskiy commented:
@Scottmitch thanks a lots, addressed your comments.

Scottmitch commented on handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java:
```diff
@@ -0,0 +1,216 @@
+                            return;
+                        }
+                        readSawAfterWrite = false;
+                        if (timeoutNanos > 0) {
```
me:

> what does this handler do if `timeoutNanos <= 0`? Should we even allow this?

@Spikhalskiy:

> ```
if (timeout <= 0) {
            timeoutNanos = 0;
        } else {
            timeoutNanos = Math.max(unit.toNanos(timeout), MIN_TIMEOUT_NANOS);
        }
```

> This is done in a consistent manner with the code in other handlers in ```io.netty.handler.timeout```

Yes I see the guard against negative in the constructor, but it doesn't look like the handler does anything meaningful if `timeout <= 0` (has this code changed or am I missing something?). I don't think there is much value in having a `ReadTimeoutHandler` with a read timeout `<=0` either so I wouldn't worry about being consistent here.

Spikhalskiy replied:
now throw IllegalArgumentException if timeoutNanos <= 0.

profic commented:
Hi, are there any reasons to not merge this PR?
It seems to be very useful feature for writing clients using Netty
Frameworks like Finagle can rid of additional threads to manage timeouts.

normanmaurer commented:
Closing as the discussion is not "active" anymore

Spikhalskiy commented:
@normanmaurer to be honest, it was not "active" from your side. I addressed all review comments and it still was abandoned and never merged without clear reasons for me. I still think that this code is very useful and we use it in production in pretty high load applications for years already.

johnou commented:
@Spikhalskiy I think you missed the questions in this thread https://github.com/netty/netty/pull/5484/files#r78883547

Spikhalskiy commented:
@johnou It's not really a question. I explained my design and explained why I made a decision that shouldn't put this functionality in the handler. Netty is a framework and a piece of functionality that @Scottmitch talks about in the comment is already implemented in another existing handler. This handler shouldn't enforce this functionality - this way we keep things 1) decoupled 2) orthogonal which gives the most flexibility to users of the framework. So, basically, we need more opinions on how it should work from you for example. If we decide that it should enforce a functionality of IdleTimeoutHandler and be less flexible - no problem, I will add it here (while I would prefer to at least write it in a way that will allow people to disable it).

Spikhalskiy commented:
@normanmaurer Could we please either reopen this PR back or just state another reason for closing if you don't want to see this functionality in netty core? It's not abandoned from my side, we actively use it and I'm ready to address the last big comment after getting a second opinion for the design point. If you just don't believe it should be in netty - let me know, I will publish it as a separate small project just in case somebody needs it because I believe that half of the projects have to implement something like this.
